tests/submodule/add.c

```diff
@@ -128,3 +128,51 @@ void test_submodule_add__url_relative_to_workdir(void)
 
 	assert_submodule_url("TestGitRepository", git_repository_workdir(g_repo));
 }
+
+void test_submodule_add__path_exists_in_index(void)
+{
+	git_index *index;
+	git_submodule *sm;
+	git_buf dirname = GIT_BUF_INIT;
+	git_buf filename = GIT_BUF_INIT;
+
+	/* In this repo, HEAD (master) has no remote tracking branc h*/
+	g_repo = cl_git_sandbox_init("testrepo");
+
+	cl_git_pass(git_buf_joinpath(&dirname, git_repository_workdir(g_repo), "TestGitRepository"));
+	cl_git_pass(git_buf_joinpath(&filename, dirname.ptr, "test.txt"));
+
+	cl_git_pass(p_mkdir(dirname.ptr, 0700));
+	cl_git_mkfile(filename.ptr, "This is some content");
+
+	cl_git_pass(git_repository_index__weakptr(&index, g_repo));
+	cl_git_pass(git_index_add_bypath(index, "TestGitRepository/test.txt"));
+
+	cl_git_fail_with(git_submodule_add_setup(&sm, g_repo, "./", "TestGitRepository", 1), GIT_EEXISTS);
+
+	git_submodule_free(sm);
+	git_buf_free(&dirname);
+	git_buf_free(&filename);
+}
+
+void test_submodule_add__file_exists_in_index(void)
+{
+	git_index *index;
+	git_submodule *sm;
+	git_buf name = GIT_BUF_INIT;
+
+	/* In this repo, HEAD (master) has no remote tracking branc h*/
+	g_repo = cl_git_sandbox_init("testrepo");
+
+	cl_git_pass(git_buf_joinpath(&name, git_repository_workdir(g_repo), "TestGitRepository"));
+
+	cl_git_mkfile(name.ptr, "Test content");
+
+	cl_git_pass(git_repository_index__weakptr(&index, g_repo));
+	cl_git_pass(git_index_add_bypath(index, "TestGitRepository"));
+
+	cl_git_fail_with(git_submodule_add_setup(&sm, g_repo, "./", "TestGitRepository", 1), GIT_EEXISTS);
+
+	git_submodule_free(sm);
+	git_buf_free(&name);
+}
```

Comment:
Right now you're checking two out of four cases:

1. an index entry exists inside of the directory you want to add
2. a file exists in the working directory

But there are two cases missing:

1. an index entry exists directly where you want to create the submodule
2. a directory exists where you want to create the submodule

Would it make sense to add tests for those two cases?

Reply:
I am not sure exactly what your test cases mean. I was trying to test two cases:

 1) A file path exists in the index with the same name as the submodule
 2) A directory exists in the index with the same name as the submodule

If you could offer some clarification I would be more than happy to add extra test cases :)

Reply:
Do you mean we should test against the workdir as well? I am not sure that my added code does that, it is an explicit check against the index but it makes sense that we should check untracked files in the workdir too. Probably a good check to have in `can_add_submodule`. Thoughts?